meetings/working-groups/extensions/anonymous-extension-declarations.md

```diff
@@ -0,0 +1,368 @@
+* People will, however, expect lookup to work very closely similar between old and new extension methods.
+
+Because the proposed design keeps using non-generic static classes as the overarching grouping entity, new extension declarations can sit side by side with old extension methods and non-extension static members. However, new extension methods do not work exactly the same as old ones. (Although do see the "Alternatives" section).
+
+### Type parameters and underlying types
+
+* Underlying types belong together with their type parameter declarations, and it would be confusing to separate them.
+* People will resent the verbosity of having to repeat underlying types and accompanying type parameters for each member.
+* Sometimes there's a need to specify different parameter details (ref-ness, nullability, ...) for the same underlying type.
+
+The design uses anonymous extension declarations within static classes. These anonymous extension declarations specify type parameters, underlying type and any parameter details for the extension member declarations they contain.
+
+### Member declarations
+
+* Extension member declarations should look identical to corresponding member declarations in classes and structs.
+* Parameter names for underlying values aren't important and people will resent the forced verbosity of having to specify them.
+
+The design keeps extension members with the same syntax as their class and struct counterparts, including using `this` to refer to the underlying value.
+
+### Regularity
+
+* Rules and syntax should be consistent across all different member kinds, generic arities etc.
+* Implementation limitations or compatibility constraints shouldn't be materially felt in the user level feature experience.
+* There should be just one right way to declare a given extension member.
+
+The design has exactly one syntactic location for each of the extension type name, underlying types with accompanying type parameters, additional parameter info and the extension members themselves.
+
+### Efficiency
+
+* Extension declarations shouldn't result in an unnecessarily large number of types being generated.
+* Extension member invocations should not incur hidden penalties, allocations or copying.
+
+The design represents anonymous extension declaration as nested static classes which are merged as much as generic arity and constraints allow. Extension members are represented as static members with an extra parameter for the underlying value, using any additional parameter info in that parameter's declaration. Extension member invocation is just invocation of those static members.
+
+## Detailed design
+
+### Declaration
+Extensions are declared inside top-level non-generic static classes, just like extension methods today, and can thus coexist with classic extension methods and non-extension static members:
+
+``` c#
+public static class Enumerable
+{
+    // New extension declaration
+    extension(IEnumerable) { ... }
```

Comment:
I like not having to specify a parameter name. There are times it could be useful, though.

Example 1. The second parameter could have `[NotNullIfNotNull(nameof(receiverParameter))]`. While `nameof(this)` _could_ be made to work there, I'm not sure it's as clear what `nameof(this)` is doing.

Example 2. If there is any chance of the IDE showing the parameter, I'd like to be able to place XML docs on it.

```cs
/// <typeparam name="T">...</typeParam>
/// <param name="receiverParamName">Deduplicated message #1</param>
extension<T>(T receiverParamName)
{
    public void A() { }
    public void B() { }
}

/// <typeparam name="T">...</typeParam>
/// <param name="receiverParamName">Deduplicated message #2</param>
extension<T>(T receiverParamName)
{
    public void D() { }
    public void E() { }
    
    // This member inherits the docs for `T`, but does not inherit the docs for `receiverParamName`:
    /// <param name="receiverParamName">Message #3</param>
    public void F() { }
}
```